// poku/lib/configs/poku.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBAL = exports.deepOptions = exports.VERSION = exports.results = exports.timespan = exports.states = void 0;
const node_process_1 = require("process");
const get_runtime_js_1 = require("../parsers/get-runtime.js");
const reporter_js_1 = require("../services/reporter.js");
exports.states = Object.create(null);
exports.timespan = Object.create(null);
exports.results = {
    passed: 0,
    failed: 0,
    skipped: 0,
    todo: 0,
};
exports.VERSION = '3.0.2';
exports.deepOptions = [];
exports.GLOBAL = {
    cwd: (0, node_process_1.cwd)(),
    configs: Object.create(null),
    configFile: undefined,
    configsFromFile: Object.create(null),
    reporter: reporter_js_1.reporter[node_process_1.env.POKU_REPORTER || 'poku'](),
    isPoku: typeof node_process_1.env.POKU_FILE === 'string' && node_process_1.env.POKU_FILE.length > 0,
    FILE: node_process_1.env.POKU_FILE,
    envFile: undefined,
    runtime: (node_process_1.env.POKU_RUNTIME || (0, get_runtime_js_1.getRuntime)()),
    runAsOnly: false,
};
